tsdapiclient/sync.py

```diff
@@ -1,43 +1,239 @@
 
-"""
-Alternatives:
-
-dir upload
-dir upload, with resume
-dir upload, with resume (client cache)
-dir upload, with mtime verification
-
-"""
-
-import time
 import os
+import time
+import sqlite3
+import sys
+
+from contextlib import contextmanager
 
+import click
+import humanfriendly.tables
 import requests
 
 from tsdapiclient.client_config import CHUNK_THRESHOLD, CHUNK_SIZE
 from tsdapiclient.fileapi import streamfile, initiate_resumable
+from tsdapiclient.tools import debug_step, get_data_path
+
+
+@contextmanager
+def sqlite_session(engine):
+    session = engine.cursor()
+    try:
+        yield session
+        session.close()
+    except Exception as e:
+        session.close()
+        engine.rollback()
+        raise e
+    finally:
+        session.close()
+        engine.commit()
+
+
+class CacheConnectionError(Exception):
+    pass
+
+
+class CacheCreationError(Exception):
+    pass
+
+
+class CacheDuplicateItemError(Exception):
+    pass
+
+
+class CacheDestroyError(Exception):
+    pass
+
+
+class CacheExistenceError(Exception):
+    pass
+
+
+class CacheItemTypeError(Exception):
+    pass
+
+
+class GenericRequestCache(object):
+
+    """sqlite-backed request cache."""
+
+    dbname = 'generic-request-cache.db'
+
+    def __init__(self):
+        self.path = f'{get_data_path()}/{self.dbname}'
```

Comment:
I think it'd be nice to either set us up a tacl subfolder, or prefix the dbname with tacl if we don't want a folder for it.

Reply:
sure.

Reply:
Now in a `/tacl/pnum` folder.